### ISSUE TITLE
Incorrect Kepler <-> Cartesian conversions?

### ISSUE THREAD

**Issue by mjuric** [ben-cassese/jorbit#26]
I tried running:

```
p = Particle.from_horizons(name="J35M00J", time=Time(61000.0, format="mjd", scale="tdb"))
p.keplerian_state
```

which gives:

```
KeplerianState(semi=Array([2.75126863], dtype=float64), ecc=Array([0.10606047], dtype=float64), inc=Array([6.91019815], dtype=float64), Omega=Array([114.83190949], dtype=float64), omega=Array([258.86133418], dtype=float64), nu=Array([216.28844803], dtype=float64), acceleration_func_kwargs={'c2': 29979.063823897617}, time=np.float64(2461000.5))
```

But that's appreciably different from the SBDB values (the epochs are the same, so I'd expect a match to a 8-9 decimal places for quantities like the semimajor axis):

<img width="894" height="541" alt="Image" src="https://github.com/user-attachments/assets/3de53649-fcf2-4559-8f4d-13d8ca8485ca" />

Is it possible there's an issue with CART->KEP conversion?

In the other direction, I've also tried constructing a `KeplerianState` from orbital elements following the [instructions here](https://jorbit.readthedocs.io/en/latest/tutorials/generate_particle_ephemeris.html); those gave ephemerides about a ~quarter of a degree away from where Horizons puts the object. KEP->CART problems as well?

Things seem to work well (match Horizons) when only cartesian states are used (which I understand `from_horizons` does under the hood.

**Comment by ben-cassese** [comment]
Hi again @mjuric, again, thanks for opening this!

I've taken a look, and the issue seems to be a misinterpretation of jorbit's output, not with the Cartesian to Keplerian transformation. Jorbit computes all state vectors, Cartesian and Keplerian, in _barycentric_ reference frames, not _heliocentric_ reference frames. For example, here is a comparison between using jorbit's `Particle.from_horizons` to create a Cartesian state vector vs. using Astroquery:

```python
import jax
jax.config.update("jax_enable_x64", True)
import jax.numpy as jnp

import numpy as np
from astropy.time import Time
import astropy.units as u

from astroquery.jplhorizons import Horizons

from jorbit import Particle
from jorbit import Ephemeris
from jorbit.utils.states import CartesianState
from jorbit.astrometry.transformations import icrs_to_horizons_ecliptic
from jorbit.utils.kepler import M_from_f


obj = Horizons(
    id="J35M00J",
    location="500@0", # note: barycentric (500@0), not heliocentric (500@10)
    epochs=Time(61000.0, format="mjd", scale="tdb").tdb.jd,
)
vecs = obj.vectors(refplane="earth")
horizons_pos = jnp.array([vecs["x"], vecs["y"], vecs["z"]]).T

p = Particle.from_horizons(
    name="J35M00J", time=Time(61000.0, format="mjd", scale="tdb")
)

jnp.min(jnp.abs(p.cartesian_state.x - horizons_pos))
```

This produces:
```
Array(0., dtype=float64)
```

Showing that they agree perfectly (which isn't surprising, since jorbit uses astroquery under the hood for these small queries to Horizons). However, when we compare the Horizons computation for the orbital elements you provided from the SBDB, they disagree:

```python
# copied over from the SBDB query
# e,a,q,i,node,peri,M
sbdb_elements = jnp.array(
    [
        0.1074501255145,
        2.740571289096321,
        2.446096560101486,
        6.920487709866278,
        114.842464101179,
        259.5145216564612,
        223.2683619260214,
    ]
)

# barycentric coordinates
obj = Horizons(
    id="J35M00J",
    location="500@0",
    epochs=Time(61000.0, format="mjd", scale="tdb").tdb.jd,
)
barycentric_elements = obj.elements(refplane="ecliptic")
barycentric_elements = jnp.array(
    [
        barycentric_elements["e"],
        barycentric_elements["a"],
        barycentric_elements["q"],
        barycentric_elements["incl"],
        barycentric_elements["Omega"],
        barycentric_elements["w"],
        barycentric_elements["M"],
    ]
).T
barycentric_elements - sbdb_elements
```

```
Array([[-0.00138965,  0.01069734,  0.01337122, -0.01028956, -0.01055461,
        -0.65318741,  0.69759753]], dtype=float64)
```

So without checking the Cartesian-Keplerian transformation we already wouldn't expect jorbit's outputs to match the SBDB, since it's starting with a slightly different state vector.

When we instead ask Horizons for the heliocentric elements, it agrees much better with the SBDB:
```python
# heliocentric coordinates
obj = Horizons(
    id="J35M00J",
    location="500@10", # now 10, not 0
    epochs=Time(61000.0, format="mjd", scale="tdb").tdb.jd,
)
heliocentric_elements = obj.elements(refplane="ecliptic")
heliocentric_elements = jnp.array(
    [
        heliocentric_elements["e"],
        heliocentric_elements["a"],
        heliocentric_elements["q"],
        heliocentric_elements["incl"],
        heliocentric_elements["Omega"],
        heliocentric_elements["w"],
        heliocentric_elements["M"],
    ]
).T
heliocentric_elements - sbdb_elements
```
```
Array([[ 3.59519359e-12, -1.57402225e-09, -1.41474610e-09,
         1.76453518e-09,  3.55760221e-09, -3.15456532e-08,
         3.27852689e-07]], dtype=float64)
```
This suggests that if we can get jorbit's initial state vector into the heliocentric frame, we should be able to recreate the SBDB elements. To do this, we need the state of the Sun at this epoch:

```python
eph = Ephemeris(
    ssos="default planets",
    earliest_time=Time("2025-01-01"),
    latest_time=Time("2026-01-01"),
)
sun_state = eph.state(Time(61000.0, format="mjd", scale="tdb"))["sun"]
sun_state
```
```
{'x': <Quantity [-0.00336382, -0.00513068, -0.00207481] AU>,
 'v': <Quantity [ 7.28782466e-06, -1.50149499e-07, -2.14655923e-07] AU / d>,
 'log_gm': Array(-8.12544774, dtype=float64, weak_type=True)}
```

Now we can subtract this off from the particle's state:
```python
p = Particle.from_horizons(
    name="J35M00J", time=Time(61000.0, format="mjd", scale="tdb")
)
bary_x = p.cartesian_state.x
bary_v = p.cartesian_state.v

helio_x = bary_x - sun_state["x"].value
helio_v = bary_v - sun_state["v"].value

```
While it would be lovely to simply create a `CartesianState` from these and convert that to a `KeplerianState`, that unfortunately won't work. Indeed:

```python
c = CartesianState(x=helio_x, v=helio_v, acceleration_func_kwargs={})
kep = c.to_keplerian()

wrong_jorbit_elements = jnp.array([
    kep.ecc,
    kep.semi,
    kep.semi * (1 - kep.ecc),
    kep.inc,
    kep.Omega,
    kep.omega,
    M_from_f(kep.nu * jnp.pi / 180, kep.ecc) * 180 / jnp.pi,
]).T[0]

wrong_jorbit_elements - sbdb_elements
```
```
Array([ 9.48331499e-04, -3.10558625e-03, -5.36791557e-03,  1.76452630e-09,
        3.55755958e-09,  4.12378411e-01, -4.15205177e-01], dtype=float64)
```

These discrepancies are is because the `to_keplerian` method relies on  the [`jorbit.astrometry.transformations.cartesian_to_elements`](https://jorbit.readthedocs.io/en/latest/core/astrometry/transformations.html#jorbit.astrometry.transformations.cartesian_to_elements) method, which by default uses the total GM of all massive objects in DE440 when computing a and e, not the GM of the Sun (both values are in [`jorbit.data.constants`](https://jorbit.readthedocs.io/en/latest/core/data/constants.html#jorbit.data.constants.TOTAL_SOLAR_SYSTEM_GM), 0.00029631 vs 0.00029591 with G in AU^3 / day^2). But, to verify that all of the transformations are working correctly and that we can recreate SBDB elements, we can see what happens if we swap in the GM of the Sun vs. the whole solar system. Below I've simply copied/pasted `jorbit.astrometry.transformations.cartesian_to_elements` and replaced TOTAL_SOLAR_SYSTEM_GM with SUN_GM




```python
# usually the conversion from the icrs frame to ecliptic frame is handled internally,
# but since we're exposing/modifying a lower-level interface
# (jorbit.astrometry.transformations.cartesian_to_elements), we need to do it manually
helio_x = icrs_to_horizons_ecliptic(helio_x)
helio_v = icrs_to_horizons_ecliptic(helio_v)

SUN_GM = 2.9591220828411951e-04

@jax.jit
def cartesian_to_elements(x: jnp.ndarray, v: jnp.ndarray) -> tuple:
    """Convert cartesian coordinates to orbital elements.

    Relies on the total mass of the solar system, which is assumed to be the sum of all
    GM values of the sun, planets, and 16 most massive asteroids as assumed by DE440.

    This is the inverse of elements_to_cartesian.

    Args:
        x (jnp.ndarray): Position in AU.
        v (jnp.ndarray): Velocity in AU/day.

    Returns:
        tuple:
            (a, ecc, nu, inc, Omega, omega) where a is the semi-major axis in AU,
            ecc is the eccentricity, nu is the true anomaly in degrees, inc is the
            inclination in degrees, Omega is the longitude of the ascending node in
            degrees, and omega is the argument of periapsis in degrees.
    """
    r_mag = jnp.linalg.norm(x, axis=1)
    v_mag = jnp.linalg.norm(v, axis=1)

    # Specific angular momentum
    h = jnp.cross(x, v)
    h_mag = jnp.linalg.norm(h, axis=1)

    # Eccentricity vector
    e_vec = jnp.cross(v, h) / SUN_GM - x / r_mag[:, jnp.newaxis]
    ecc = jnp.linalg.norm(e_vec, axis=1)

    # Specific orbital energy
    specific_energy = v_mag**2 / 2 - SUN_GM / r_mag

    a = -SUN_GM / (2 * specific_energy)

    inc = jnp.arccos(h[:, 2] / h_mag) * 180 / jnp.pi

    n = jnp.cross(jnp.array([0, 0, 1]), h)
    n_mag = jnp.linalg.norm(n, axis=1)

    Omega = jnp.where(
        n[:, 1] >= 0,
        jnp.arccos(n[:, 0] / n_mag) * 180 / jnp.pi,
        360.0 - jnp.arccos(n[:, 0] / n_mag) * 180 / jnp.pi,
    )
    Omega = jnp.where(n_mag == 0, 0, Omega)

    omega = jnp.where(
        n_mag > 0,
        jnp.where(
            e_vec[:, 2] >= 0,
            jnp.arccos(
                jnp.clip(
                    jnp.sum(n * e_vec, axis=1)
                    / (n_mag * jnp.linalg.norm(e_vec, axis=1)),
                    -1,
                    1,
                )
            )
            * 180
            / jnp.pi,
            360
            - jnp.arccos(
                jnp.clip(
                    jnp.sum(n * e_vec, axis=1)
                    / (n_mag * jnp.linalg.norm(e_vec, axis=1)),
                    -1,
                    1,
                )
            )
            * 180
            / jnp.pi,
        ),
        0,
    )

    nu = jnp.where(
        jnp.sum(x * v, axis=1) >= 0,
        jnp.arccos(jnp.clip(jnp.sum(e_vec * x, axis=1) / (ecc * r_mag), -1, 1))
        * 180
        / jnp.pi,
        360
        - jnp.arccos(jnp.clip(jnp.sum(e_vec * x, axis=1) / (ecc * r_mag), -1, 1))
        * 180
        / jnp.pi,
    )

    return a, ecc, nu, inc, Omega, omega

_a, _ecc, _nu, _inc, _Omega, _omega = cartesian_to_elements(helio_x, helio_v)
_M = M_from_f(_nu*jnp.pi/180, _ecc) * 180 / jnp.pi
_q = _a*(1-_ecc)

new_elements = jnp.array([_ecc, _a, _q, _inc, _Omega, _omega, _M]).T[0]
new_elements - sbdb_elements
```
```
Array([ 3.59520747e-12, -1.57402269e-09, -1.41474610e-09,  1.76452630e-09,
        3.55755958e-09, -3.15454827e-08,  3.27852575e-07], dtype=float64)
```
So, once we create a CartesianState in jorbit in the heliocentric frame and use just the Sun's GM, we can recreate the SBDB orbital elements to your expected high precision.

Sorry for such a long answer! Let me know if this all makes sense, or if it would be helpful to create/expose an interface that does the transformation automatically. I can imagine adding a `.to_heliocentric` method to `KeplerianState`, but let me know if that's worth doing.
